README.md
# Graph Analysis

Graph analysis adds the **analysis view** to Obsidian which implements a set of
algorithms that computes useful relations between the notes in your vault! Our
flagship algorithm is the **Co-citations** panel, that we describe as a _2nd
order backlinks panel_.

The Graph Analysis view shows a table of note names and numbers, each
representing the value of some graph analysis algorithm on that note in relation
to the current note.

e.g.

- `[[A]] is 0.9 Similar to [[B]]`
- `[[A]] has a 0.6 chance of being connected to [[B]]`
- `[[A]] is co-cited with [[B]] 6 times`

## Analysis Types

Graph Analysis currently has 3 different analysis types:

1. Similarity
2. Link Prediction
3. Co-Citations
4. Community Detection

Each implement different algorithms with different purposes.

### Co-Citations

Co-Citations counts the number of time two notes are cited together in the same
note and gives extra weight when the two notes are cited close together.

Think of co-citations as a **2nd-order backlinks** panel: Instead of showing
_where_ something is cited, it shows _why_, or with _whom_ or _what_ it is
cited!

For example, if `[[C]]` has `[[A]] and [[B]]` in its content, then `[[A]]` and
`[[B]]` will each have a co-citation of one.

Each note with co-citations > 0 is given a drop down menu. Inside each drop
down, you can see which note co-cites those two notes, and the sentence in which
they are co-cited (if in the same sentence), otherwise just the sentence with
the other link.

![](https://i.imgur.com/9yspOkN.png)

#### Example use case with daily notes

An example why this is useful is given by @HEmile:

> I use a lot of daily notes, in which I journal and write about the news of the
> day. This makes the backlinks panel a bit boring: It only shows on what dates
> I wrote about some note. The Co-Citations algorithm shows me much more! For
> example, the `Joe Biden` note shows me I usually write about Biden together
> with `Donald Trump`. But if I want to know what I wrote about the relations
> between Joe Biden and `China`, I can just look in the co-citations panel and
> expand the relation to see the story!

![](https://i.imgur.com/udPkuV3.png)

#### Video Tutorial
This video gives a longer and in depth overview for why Co-Citations is so useful!
[![Watch the video](https://yt-embed.herokuapp.com/embed?v=rK6JVDrGERA)](https://youtu.be/rK6JVDrGERA)

### Similarity

Similarity is a measure of how similar two notes are based on their
connectedness in the graph (ie. note content is not considered). Currently, only
the Jaccard Similarity measure is implemented.

#### Jaccard Similarity

**Formula**:

![image](https://user-images.githubusercontent.com/70717676/139872572-93504295-6d29-4722-bdb1-3fbeb7bc22ec.png)

[Source](https://neo4j.com/docs/graph-data-science/current/alpha-algorithms/jaccard/#alpha-algorithms-similarity-jaccard-context)

Where

- `|x|` is the number of neighbours the node `x` has (links going in or out).
- `|x & y|` is the number of neighbours that both `x` and `y` have in common

### Link Prediction

Link Prediction is a measure of the probability that two notes should be
connected based on their other connections in the graph. The implemented Link
Prediction algorithms are Adamic Adar and Common Neighbours.

#### Adamic Adar

**Formula**:

![image](https://user-images.githubusercontent.com/70717676/139873180-c870e072-843c-42a9-83fc-87205b408754.png)

[Source](https://neo4j.com/docs/graph-data-science/current/alpha-algorithms/adamic-adar/)

Where:

- `N(x)` is the number of neighbours of `x`

#### Common Neighbours

**Formula**:

![image](https://user-images.githubusercontent.com/70717676/139873406-d0542335-3b8c-4d08-8a5b-4510408ebd4e.png)

[Source](https://neo4j.com/docs/graph-data-science/current/alpha-algorithms/common-neighbors/)

Where:

- `N(x)` is the numbers of neighbours of `x`

### Community Detection

These algorithms try to find groups of similar notes.

#### Label Propagation

Start by giving each node a unique label (its own name). Then, look at each node's neighbours, and change it's label to the most common among it's neighbours. 
Repeat this process `iterations` number of times. 

At the end, show the nodes grouped by the last label they had.

#### Clustering Coefficient

Gives the ratio of the number of _triangles_ the `u` is a part of, to the number of triangles it possibly _could have_ been a part of:

![image](https://user-images.githubusercontent.com/70717676/140610147-0a05201f-d9c7-4c0c-b423-6bbeeb81253b.png)

## Utility Classes

Each row in the graph analysis tables (or co-citations dropdowns) has a class:
`analysis-linked` or `analysis-not-linked`, indicating if the current note is
linked to the note in that row. This gives you the ability to style a table row
based on whether it's connectedness.

For example, you can make linked notes have a lower opacity:

```css
tr.analysis-linked {
  opacity: 0.3;
}
```

![image](https://user-images.githubusercontent.com/70717676/139862955-75284ff5-0ced-4548-bf6e-caa353a16fe0.png)

You could even go so far as to hide linked rows completely:

```css
tr.analysis-linked {
  display: none;
}
```

## Settings

In the analysis view, you have the option to choose between different
`Analysis Types`, and different `Algorithms` within those types. You can set the
default analysis type in the plugin settings.

There is also the option to hide `Infinity` and `Zero` values.

![image](https://user-images.githubusercontent.com/70717676/138652879-d8b0e4a7-d70a-44e8-ba3c-67e04f6a8edd.png)

## Documentation on Algorithms

You can read more about the implemented algorithms, or let us know which you
want us to add, over
[here](https://neo4j.com/docs/graph-data-science/current/algorithms/) 👀.
Information on co-citations can mostly be found on
[Wikipedia](https://en.wikipedia.org/wiki/Co-citation). In particular, we
implement a variation of
[Co-citatition Proximity Analysis](https://en.wikipedia.org/wiki/Co-citation_Proximity_Analysis).


## Buy Us a Coffee

SkepticMystic: [![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/G2G454TZF)

Emile: [![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/Emile)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.15.4](https://github.com/SkepticMystic/graph-analysis/compare/0.15.3...0.15.4) (2022-01-31)

### [0.15.3](https://github.com/SkepticMystic/graph-analysis/compare/0.15.2...0.15.3) (2022-01-12)


### Features

* Now hides algorithm selector if only 1 alg is active ([5391a47](https://github.com/SkepticMystic/graph-analysis/commit/5391a47971df2490de635465abbfb8c65d93922c))


### Bug Fixes

* Better styling of ordered lists and footnotes ([79e9ca1](https://github.com/SkepticMystic/graph-analysis/commit/79e9ca162ac80f31e1764a7d35cef32fbba5e277))

### [0.15.2](https://github.com/SkepticMystic/graph-analysis/compare/0.15.1...0.15.2) (2022-01-10)


### Bug Fixes

* Invalid Markdown rendering in co-cites for child bullets ([4827e85](https://github.com/SkepticMystic/graph-analysis/commit/4827e8597d32f3c772596ed8980ed75da737fe96))

### [0.15.1](https://github.com/SkepticMystic/graph-analysis/compare/0.15.0...0.15.1) (2022-01-10)


### Features

* Added outliner support to co-citations ([0eef53b](https://github.com/SkepticMystic/graph-analysis/commit/0eef53b9db1611160b3bc6a6b0ec61bcfca9568c))
* Increased granularity for co-citations on the same line ([cfca1d5](https://github.com/SkepticMystic/graph-analysis/commit/cfca1d53e8e4d7674db1b8f536ff0bf540bf3f31))
* Sort equal-measure co-citations alphabetically ([62b351a](https://github.com/SkepticMystic/graph-analysis/commit/62b351a57184a8b7335960728c334aaaaec4f620))
* Text previews in Co-Citations now render markdown ([5d1153b](https://github.com/SkepticMystic/graph-analysis/commit/5d1153b366da6d19aed2a3ba2b1901fe0695081e))


### Bug Fixes

* Bug in filles without outlines ([00c4061](https://github.com/SkepticMystic/graph-analysis/commit/00c406168c81b366d51a8e4ea16fd105bcc40001))
* Bugs in sorting results of co-citation ([2012a91](https://github.com/SkepticMystic/graph-analysis/commit/2012a9191347dc724005b7e4ae579c84eba0400a))
* Results within co-citations entries were not properly sorted ([318666b](https://github.com/SkepticMystic/graph-analysis/commit/318666b0fa6c9bd2839f9e21e82ba73b92c30c67))

## [0.15.0](https://github.com/SkepticMystic/graph-analysis/compare/0.14.1...0.15.0) (2022-01-03)

### [0.14.1](https://github.com/SkepticMystic/graph-analysis/compare/0.14.0...0.14.1) (2022-01-03)


### Features

* :sparkles: Option to hide/show each alg (not all are shown by default anymore) ([242f8f9](https://github.com/SkepticMystic/graph-analysis/commit/242f8f9c5b924d04be6394edb210f22347a61e9b))

## [0.14.0](https://github.com/SkepticMystic/graph-analysis/compare/0.13.2...0.14.0) (2021-12-20)

### [0.13.2](https://github.com/SkepticMystic/graph-analysis/compare/0.13.1...0.13.2) (2021-12-20)


### Features

* **NLP:** :sparkles: Sentiment analysis global alg! ([8c5d49a](https://github.com/SkepticMystic/graph-analysis/commit/8c5d49a40214e983c28556a1905a70c77e9d351a))


### Bug Fixes

* :bug: Don't refresh a global alg ([0eb5d42](https://github.com/SkepticMystic/graph-analysis/commit/0eb5d4259b79c9e855b831f888e32e3c8e74bdb3))
* :bug: Show currNode for global algs ([d0695ed](https://github.com/SkepticMystic/graph-analysis/commit/d0695eded684766eeedd65f0974551729fdb7b44))

### [0.13.1](https://github.com/SkepticMystic/graph-analysis/compare/0.13.0...0.13.1) (2021-12-19)

## [0.13.0](https://github.com/SkepticMystic/graph-analysis/compare/0.12.0...0.13.0) (2021-12-18)


### Features

* **NLP:** :sparkles: 3 new Natural Language Processing algorithms!!! ([84fd96f](https://github.com/SkepticMystic/graph-analysis/commit/84fd96f03a53aaa433947b7b6123fe41b9bacaa6))

## [0.12.0](https://github.com/SkepticMystic/graph-analysis/compare/0.11.0...0.12.0) (2021-11-16)


### Features

* **HITS:** :sparkles: HITS Algorithm! ([46a7b91](https://github.com/SkepticMystic/graph-analysis/commit/46a7b91f035c5992bc6c8849fef829355636054c))

## [0.11.0](https://github.com/SkepticMystic/graph-analysis/compare/0.10.3...0.11.0) (2021-11-16)


### Features

* :sparkles: Sliders on labelProp and Louvain! ([6ee2223](https://github.com/SkepticMystic/graph-analysis/commit/6ee22233f0e0e4e725fd12494f6a5753e531dac8))
* :sparkles: Tell the type of analysis in the info-tooltip ([67166c6](https://github.com/SkepticMystic/graph-analysis/commit/67166c607e06041817d77e93cedbe71cc500b2e2))
* **commDetection:** :sparkles: Louvain commDetection! ([1ba9a41](https://github.com/SkepticMystic/graph-analysis/commit/1ba9a41220cab5cf36f66ef2d9c71b964f6a3df2))

### [0.10.3](https://github.com/SkepticMystic/graph-analysis/compare/0.10.2...0.10.3) (2021-11-15)


### Bug Fixes

* :bug: ExclusionTags, allow empty value (Fix [#39](https://github.com/SkepticMystic/graph-analysis/issues/39)) ([4d6004d](https://github.com/SkepticMystic/graph-analysis/commit/4d6004d8669b6dda9a4eec92b1f80000939ad6ed))

### [0.10.2](https://github.com/SkepticMystic/graph-analysis/compare/0.10.1...0.10.2) (2021-11-14)


### Features

* **coCites:** :sparkles: Select coCite.sentence when clicking it (fix [#38](https://github.com/SkepticMystic/graph-analysis/issues/38)) ([043d0b2](https://github.com/SkepticMystic/graph-analysis/commit/043d0b20927adaa7e43ba784a0bb20259f34cca5))

### [0.10.1](https://github.com/SkepticMystic/graph-analysis/compare/0.10.0...0.10.1) (2021-11-13)


### Bug Fixes

* :bug: Settings weren't loading ([b5b4a82](https://github.com/SkepticMystic/graph-analysis/commit/b5b4a82eee66b120a5997a352b1cb7499d69bd0b))

## [0.10.0](https://github.com/SkepticMystic/graph-analysis/compare/0.9.0...0.10.0) (2021-11-13)


### Features

* :sparkles: Add `.GA-${ext}` to nodes for styling ([85cbb79](https://github.com/SkepticMystic/graph-analysis/commit/85cbb79a51d56bc37120c93560ef547c305f3db6))
* :sparkles: Commands to open specific alg type ([8801c6c](https://github.com/SkepticMystic/graph-analysis/commit/8801c6c9c3a77b91b3e74b1b48192a09cd01c2b8))

src/Constants.ts
import type { GraphAnalysisSettings, SubtypeInfo } from 'src/Interfaces'

export const DEFAULT_SETTINGS: GraphAnalysisSettings = {
  noInfinity: true,
  noZero: true,
  allFileExtensions: true,
  showImgThumbnails: true,
  addUnresolved: true,
  coTags: true,
  defaultSubtypeType: 'Co-Citations',
  debugMode: false,
  superDebugMode: false,
  exclusionRegex: '',
  exclusionTags: [],
  algsToShow: [
    'Adamic Adar',
    'Jaccard',
    'Otsuka-Chiai',
    'Co-Citations',
    'Label Propagation',
  ],
}

export const DECIMALS = 4

export const VIEW_TYPE_GRAPH_ANALYSIS = 'graph-analysis'

export const LINKED = 'GA-linked'
export const NOT_LINKED = 'GA-not-linked'

export const MEASURE = 'GA-measure'
export const NODE = 'GA-node'

export const ICON = 'GA-icon'

export const ANALYSIS_TYPES: SubtypeInfo[] = [
  {
    anl: 'Co-Citations',
    subtype: 'Co-Citations',
    desc: 'See which of your notes are referenced together most often.',
    global: false,
    nlp: false,
  },
  {
    anl: 'Centrality',
    subtype: 'HITS',
    desc: 'An authority has lots of links coming in\nA hub has lots of links going out.',
    global: true,
    nlp: false,
  },
  {
    anl: 'Link Prediction',
    subtype: 'Adamic Adar',
    desc: 'Based on the structure of your graph, this alg predicts which notes _should_ be linked to the current note.',
    global: false,
    nlp: false,
  },
  // {
  //   anl: 'Link Prediction',
  //   subtype: 'Common Neighbours',
  //   desc: 'Tells you how many notes are linked to the current (active) note, and the note in the table.\nHover over a cell in the table to see a list of common neighbours',
  //   global: false,
  //   nlp: false,
  // },

  {
    anl: 'Similarity',
    subtype: 'Jaccard',
    desc: "Based on the structure of your graph, this alg predicts which notes are most similar to the current note.\n\nIt shows the ratio of the numbers of neighbours two notes have in common, to the total number of neighbours they each have.\n\n'🔗' means that this note is linked to the group name.",
    global: false,
    nlp: false,
  },
  {
    anl: 'Similarity',
    subtype: 'Overlap',
    desc: '<No description given yet>',
    global: false,
    nlp: false,
  },

  {
    anl: 'Community Detection',
    subtype: 'Label Propagation',
    desc: "Start by giving each node a unique label (its own name). Then, look at each node's neighbours, and change it's label to the most common among it's neighbours. Repeat this process `iterations` number of times. Show the nodes grouped by the last label they had.\n\n'🔗' means that this note is linked to the group name.",
    global: true,
    nlp: false,
  },
  {
    anl: 'Community Detection',
    subtype: 'Louvain',
    desc: "Show the Louvain community that the current note is in.\n\n'🔗' means that this note is linked to the group name.",
    global: false,
    nlp: false,
  },
  {
    anl: 'Community Detection',
    subtype: 'Clustering Coefficient',
    desc: 'Gives the likelihood that a nodes _neighbours_ are connected to each other.',
    global: true,
    nlp: false,
  },
  {
    anl: 'NLP',
    subtype: 'BoW',
    desc: 'Split a note into its words, count how many times each word appears, and use that to compare similarity between notes.',
    global: false,
    nlp: true,
  },
  // {
  //   anl: 'NLP',
  //   subtype: 'Tversky',
  //   desc: '',
  //   global: false,
  //   nlp: true,
  // },
  {
    anl: 'NLP',
    subtype: 'Otsuka-Chiai',
    desc: 'Returns the "Otsuka-Chiai" similarity between the current note and every other note.',
    global: false,
    nlp: true,
  },
  {
    anl: 'NLP',
    subtype: 'Sentiment',
    desc: 'Gives the sentiment of every note. Positive → higher sentiment.',
    global: true,
    nlp: true,
  },
]

export const IMG_EXTENSIONS = ['png', 'jpg', 'jpeg', 'svg', 'gif', 'webp']

export const iconSVG = `<path fill="currentColor" stroke="currentColor" d="M88.8,67.5c-3,0-5.7,1.2-7.7,3.1l-12.2-7c0.7-1.9,1.2-3.9,1.2-6.1C70,47.8,62.2,40,52.5,40c-1.3,0-2.6,0.2-3.8,0.5l-5-10.8
c2.3-2.1,3.8-5,3.8-8.4c0-6.2-5-11.3-11.3-11.3S25,15,25,21.3s5,11.3,11.3,11.3c0.1,0,0.3,0,0.4,0l5.2,11.2
c-4.2,3.2-6.9,8.2-6.9,13.8C35,67.2,42.8,75,52.5,75c4.8,0,9.2-1.9,12.3-5.1l12.8,7.3c-0.1,0.5-0.2,1-0.2,1.5
c0,6.2,5,11.3,11.3,11.3S100,85,100,78.7S95,67.5,88.8,67.5z M36.3,25c-2.1,0-3.8-1.7-3.8-3.8s1.7-3.8,3.8-3.8s3.8,1.7,3.8,3.8
S38.3,25,36.3,25z M52.5,67.5c-5.5,0-10-4.5-10-10s4.5-10,10-10s10,4.5,10,10S58,67.5,52.5,67.5z M88.8,82.5c-2.1,0-3.8-1.7-3.8-3.8
s1.7-3.8,3.8-3.8s3.8,1.7,3.8,3.8S90.8,82.5,88.8,82.5z M80.3,41.7l-3-4l-7.5,5.6l3,4L80.3,41.7z M90,40c5.5,0,10-4.5,10-10
s-4.5-10-10-10s-10,4.5-10,10S84.5,40,90,40z M23.8,60h7.5v-5h-7.5V60z M10,47.5c-5.5,0-10,4.5-10,10s4.5,10,10,10s10-4.5,10-10
S15.5,47.5,10,47.5z"/>`
